pkg/pmem-csi-operator/controller/deployment/deployment_controller.go

```diff
@@ -47,25 +53,128 @@ func Add(mgr manager.Manager, opts pmemcontroller.ControllerOptions) error {
 	if err != nil {
 		return err
 	}
-	return add(mgr, r)
+	return add(mgr, r.(*ReconcileDeployment))
 }
 
 // add adds a new Controller to mgr with r as the reconcile.Reconciler
-func add(mgr manager.Manager, r reconcile.Reconciler) error {
+func add(mgr manager.Manager, r *ReconcileDeployment) error {
 	// Create a new controller
 	c, err := controller.New("deployment-controller", mgr, controller.Options{Reconciler: r})
 	if err != nil {
 		klog.Errorf("Deployment.Add: error: %v", err)
 		return err
 	}
 
+	p := predicate.Funcs{
+		UpdateFunc: func(e event.UpdateEvent) bool {
+			r.reconcileMutex.Lock()
+			defer r.reconcileMutex.Unlock()
+			klog.Infof("UPDATED: %T/%s, generation:%d", e.MetaOld, e.MetaOld.GetName(), e.MetaNew.GetGeneration())
+			old := e.ObjectOld.(*pmemcsiv1alpha1.Deployment)
+			new := e.ObjectNew.(*pmemcsiv1alpha1.Deployment)
+			if new.DeletionTimestamp != nil {
+				// Deployment CR deleted, remove it's reference from cache.
+				// Objects owned by it are automatically garbage collected.
+				r.deleteDeployment(e.MetaOld.GetName())
+				return false
+			}
+			if old.GetGeneration() == new.GetGeneration() {
+				// No changes registered
+				return false
+			}
+			// We are intersted in only spec changes, not CR status changes
+			changes := old.Compare(new)
```

Comment:
Below are points that led me to keep compare:
 - This makes it easy to understand by looking at logs what exactly changed that leads to "deployment reconcile"
 - "Update" gets invoked on each time status got updated, which we clearly can avoid reconciling. Otherwise, the logs look messy with quite many "redeploy...." statements.

Reply:
So basically we maintain code that we need to test for correctness just for logging.

Can we use some automated diff and print the result of that instead of manually writing the code? There's https://pkg.go.dev/github.com/google/go-cmp/cmp specifically for that purpose or we can generate a patch and print that (probably less readable, but avoids the extra dependency).

Reply:
Replaced `Compare()` with  `Patch.Data()`, which does the json patching for computing the changes,.